### PR TITLE
Tracker module for ML based TC tracking algortihms

### LINKED ISSUE #200: Locate the storm centres in the ML detection
The ML neural network will output values / probabilities for each of the tropical cyclone categories at each grid point. We will want to translate this to the centre points of the storms. This can be done by filtering to find all the grid points above a certain probability and then finding the centroid. This is complicated by the fact that there may be multiple TCs at a given time, so there will need to be some way to separate out multiple storms

### PR DESCRIPTION
**Pull request by surbhigoel77**
Linked issue: #190

 The PR aims to build the core infrastructure for running ML models within the TCTrack framework. It uses exisiting base classes and defines new ml-specific classes.

It includes the following files (based on #204):

- [ ] `ml_tracker.py` : contains a base dataclass `TCMLParameters` to fetch ML Parameters and a base abstract class `TCMLTracker` that extends the existing Tracker by adding a load_model function (for ml models). These classes remain common for all ML models (basically it is a way to standardise processes for any incoming new model).
- [ ] `cyclone_track_ml.py` : inherits the base classes defined in `ml_tracker.py` and adds new model specific functions. Reads the model from hugging face hub repository and connects it to the run_tracker pipeline. For classes that are common for both ML and non-ML models like `TCTrackerMetadata` and `Trajectory` are inherited from the original `core/tracker.py`. 

- [ ] `test_ml_tracker.py` : unit, integration

Current update:

1. Inference: preprocess(), detect() where for each timestep an inference is run and points are classified under class 0 to 4 and this information is saved for each point. This process if repeated for each timestep.
2. Clustering:  cluestering_candidates() where clustering happens of neigbouring/adjacent points with same class. These clusters are called as candidates.
3. Merging nearby candidates
4. Check for nearest unmerged candidates
5.  Track stitching: stitch() where candidates that are of interest (based on a threshold) are linked together to form a TC track using above functions.

Assumptions:

merge_distance_deg = 2.0 (How close (in degrees) two candidates in the same timestep can be before the weaker one is discarded as a duplicate)
stitch_max_distance_deg = 3.0 (How far (in degrees) a candidate can be from a track's last known position and still be claimed as that same storm continuing)
stitch_max_gap: int = 1 (How many timesteps in a row a track can go unmatched before it's retired)
stitch_min_length: int = 2 (The minimum number of points a track must have to be kept in the final output)

### CODE REVIEW

**Comment by surbhigoel77** [comment]
@sjavis you can try and run the ML Tracker using the 'run_tracker_on_real_data.py' script. You will need ERA5 data and the normalisation stats to run this. I have already provided the absolute paths/creds in the code for

1. the hugging face repo for model weights, and 
2. the required ERA5 data and norm stats

Let me have issues accessing it.

The integration test is not to be reviewed atm, I am making changes to it

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 1]
```diff
@@ -0,0 +1,837 @@
+"""Module providing a class to run a machine-learning tracking algorithm.
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py:1]: I suggest you add `MLParameters` and `MLTracker` to a `__init__.py` file so it can be accessed by `import tctrack.machine_learning`, matching the other tracking algorithms.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 2]
```diff
@@ -0,0 +1,837 @@
...
+
+    stitch_max_gap: int = 1
+    """Number of consecutive timesteps a track may go unmatched before it is
+    closed, tolerating brief drops below :attr:`~TCMLParameters.threshold`.
+    """
+
+    stitch_min_length: int = 2
+    """Minimum number of observations a track must have to be kept, filtering
+    out single-timestep detections likely to be noise.
+    """
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py:188]: I suggest making a separate class for the stitching parameters.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 3]
```diff
@@ -0,0 +1,837 @@
...
+            Class containing the parameters for the tracking algorithm.
+
+        Raises
+        ------
+        OSError
+            If ``parameters.model_path`` is given but the file does not exist.
+        huggingface_hub.errors.RepositoryNotFoundError
+            If the HuggingFace repository cannot be found or accessed.
+        """
+        self.parameters: MLParameters = parameters
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py:240]:
```suggestion
        self.parameters: MLParameters = parameters
        self._hf_token: str | None = self.parameters.hf_token
        self.parameters.hf_token = None
```
As we discussed, it would be a good idea to clear the token from the parameters class so it doesn't get distributed in the output files. Any places where it is used would need to be updated.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 4]
```diff
@@ -0,0 +1,837 @@
...
+        ):
+            standard_name, units = colocated_metadata[variable]
+            described = standard_name.replace("_", " ")
+            self._variable_metadata[f"{variable}_{int(level)}"] = TCTrackerMetadata(
+                properties={
+                    "standard_name": standard_name,
+                    "long_name": f"{described} at {int(level)} hPa",
+                    "units": units,
+                }
+            )
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py:341]: It would probably be best to allow the variables to have an additional pressure level dimension so only one field is needed per variable. But that would require changes to `to_netcdf` so it is a separate issue. This seems like a good workaround for now.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 5]
```diff
@@ -0,0 +1,837 @@
...
+        ------
+        ValueError
+            If :attr:`parameters.normalisation_stats_path` is not set.
+        """
+        if self.parameters.normalisation_stats_path is None:
+            msg = (
+                "parameters.normalisation_stats_path must be set to normalise "
+                "model input - see MLTracker.preprocess."
+            )
+            raise ValueError(msg)
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py]: As discussed, I suggest including the stats file with the code. Then that can be used if `normalisation_stats_path` is None.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 6]
```diff
@@ -0,0 +1,837 @@
...
+                for i, lat in enumerate(self._lats):
+                    for j, lon in enumerate(self._lons):
+                        self._scores.append( #scores list collects the time, lat, lon and the predicted class probabilities for each pixel in the input frame.
+                            {
+                                "time": self._times[t],
+                                "lat": float(lat),
+                                "lon": float(lon),
+                                "probs": probs[:, i, j].tolist(),
+                            }
+                        )
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py:545]: `_scores` seems like it isn't currently used anywhere. Is that right? Is there an intention to do something with it?

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 7]
```diff
@@ -0,0 +1,837 @@
...
+            
+            # Run the loop until no neigbouring pixel is left unchecked for a blob
+            while queue:
+                y, x = queue.popleft() # popleft() removes and returns an element from the left end of the deque
+                for dy, dx in ((-1, 0), (1, 0), (0, -1), (0, 1)):
+                    neighbour = (y + dy, x + dx)
+                    if (
+                        neighbour in unvisited
+                        and class_idx[neighbour] == component_class
+                    ):
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py]: I can imagine a (possibly common) situation in which different pixels may be classed differently for a single storm. Under this approach it would identify it as two storms. Perhaps it should not care about class for this stage of the clustering. Then the class index can be determined from the list of pixels later.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 8]
```diff
@@ -0,0 +1,837 @@
...
+            Winning class probability per pixel, shape ``(lat, lon)``.
+
+        Returns
+        -------
+        list[dict]
+            One dict per cluster with keys ``lat``, ``lon``, ``class_index``,
+            and ``score`` - the probability-weighted centroid, class, and
+            peak probability of each connected group of same-class pixels.
+        """
+        unvisited = set(zip(*np.nonzero(is_storm), strict=False)) #storm classified pixels that are still unclaimed by any blob.
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py]:
```suggestion
        unvisited = set((lat, lon) for lat, lon in np.argwhere(is_storm)) #indices of storm classified pixels that are still unclaimed by any blob.
```
I think this is a little clearer way of doing the same thing

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 9]
```diff
@@ -0,0 +1,837 @@
...
+        OSError
+            If ``parameters.model_path`` is given but the file does not exist.
+        huggingface_hub.errors.RepositoryNotFoundError
+            If the HuggingFace repository cannot be found or accessed.
+        """
+        self.parameters: MLParameters = parameters
+        self._trajectories: list[Trajectory] = []
+        self._scores: list[dict] = []
+        # TC locations found by detect(), consumed by stitch().
+        self._candidates: list[dict] = []
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py]: I suggest we make a candidate class (maybe using a TypedDict) since it is used in a few places and it would make the typing clearer than just using a dict.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 10]
```diff
@@ -0,0 +1,837 @@
...
+        """
+        if self.parameters.merge_distance_deg <= 0: 
+            return candidates
+
+        kept: list[dict] = [] #list that will hold the strongest candidates after eliminating the weaker ones that are too close to each other.
+        for candidate in sorted(candidates, key=lambda c: c["score"], reverse=True): #
+            if all(
+                np.hypot(#calculate distance between centroids of different candidates.
+                    candidate["lat"] - other["lat"], candidate["lon"] - other["lon"] 
+                )
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py]: This won't work for when the longitude crosses the date line. Options I see for this:
- Use the Haversine function. This would be the most clear and accurate, bu may be overkill since near the poles don't really matter. It could be useful to have the function anyway.
- Change the longitude difference to `(candidate["lon"] - other["lon"] + 180) % 360 - 180` to keep the difference below 180°. This still becomes inaccurate as you move away from the equator.
- Do the above and multiply the longitude difference by cos of the (average) latitude. This accounts for the reduction in the longitude spacing as you move away from the equator.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 11]
```diff
@@ -0,0 +1,837 @@
...
+        Returns
+        -------
+        list[dict]
+            The retained candidates, strongest first.
+        """
+        if self.parameters.merge_distance_deg <= 0: 
+            return candidates
+
+        kept: list[dict] = [] #list that will hold the strongest candidates after eliminating the weaker ones that are too close to each other.
+        for candidate in sorted(candidates, key=lambda c: c["score"], reverse=True): #
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py]:
```suggestion
        for candidate in sorted(candidates, key=lambda c: c["score"], reverse=True): # Prioritise candidates with higher confidence
```

File: run_tracker_on_real_data.py
Comment (sjavis) [inline run_tracker_on_real_data.py:1]: Just to clarify, is this file intended just for development and to be removed before merging?

**Review by sjavis — COMMENTED** [review]
I haven't reviewed the stitching or the integration test yet but I thought I would give my suggestions so far. Overall the structure seems good.

File: src/tctrack/machine_learning/cyclone_track_ml.py (hunk 6 above)
Comment (surbhigoel77) [reply src/tctrack/machine_learning/cyclone_track_ml.py:545]: It is actually kept for future per-class evaluation and also being used in the integration test. So, not dead code

File: src/tctrack/machine_learning/cyclone_track_ml.py (hunk 2 above)
Comment (sjavis) [reply src/tctrack/machine_learning/cyclone_track_ml.py:188]: This will also need adding to the `__init__.py` file

File: src/tctrack/machine_learning/cyclone_track_ml.py (hunk 4 above)
Comment (sjavis) [reply src/tctrack/machine_learning/cyclone_track_ml.py:341]: Added issue #258 to track this.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 12]
```diff
@@ -0,0 +1,925 @@
...
+    """Approximate great-circle distance between two points, in degrees.
+
+    Plain ``hypot(dlat, dlon)`` on raw coordinate differences is wrong in two
+    ways: it overstates distances that cross the antimeridian (e.g. 179E to
+    179W is 2 degrees apart, not 358), and it ignores that a degree of
+    longitude covers less physical distance away from the equator. This
+    wraps the longitude difference to ``(-180, 180]`` and scales it by the
+    cosine of the mean latitude to correct for both. Still an approximation
+    - the Haversine formula would be exact - but adequate at the latitudes
+    tropical cyclones occur at.
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py:79]:
```suggestion
    An approximation - the Haversine formula would be exact - but adequate at small
    distances and the latitudes tropical cyclones occur at.
```

A smaller description is probably sufficient.

File: src/tctrack/machine_learning/cyclone_track_ml.py (hunk 9 above)
Comment (sjavis) [reply src/tctrack/machine_learning/cyclone_track_ml.py]: This is looking good and I realise it was more complicated than I anticipated as I thought `extra_keys` was available (rather than being new in python 3.15).

I suggest making the extra variables inside a `data` dictionary to resolve the need for casting, etc. I will make a separate PR to show you my suggestion as it requires changes in a few places. (Added in #259)

File: src/tctrack/core/ml_tracker.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
+"""Module providing abstract base classes for ML-based tracking algorithms."""
+
+import os
+from abc import abstractmethod
+from dataclasses import dataclass
+
+import scipy  # noqa: F401 
+import torch
+from huggingface_hub import hf_hub_download
```
Comment (sjavis) [inline src/tctrack/core/ml_tracker.py:9]: torch and huggingface_hub should be added as dependencies to the pyproject.toml

File: src/tctrack/machine_learning/cyclone_track_ml.py (hunk 2 above)
Comment (sjavis) [reply src/tctrack/machine_learning/cyclone_track_ml.py:188]: And the parameters can probably loose the `stitch_` prefix since they are now in the class.

File: src/tctrack/machine_learning/cyclone_track_ml.py [hunk 14]
```diff
@@ -0,0 +1,925 @@
...
+            data=cf.Data(
+                date2num(
+                    [candidate["time"] for candidate in self._candidates],
+                    units=self.time_metadata["units"],
+                    calendar=self.time_metadata["calendar"],
+                ).tolist()
+            ),
+            properties={
+                "standard_name": "time",
+                "long_name": "time of detection",
```
Comment (sjavis) [inline src/tctrack/machine_learning/cyclone_track_ml.py:840]: I suggest removing the `of detection` from the long names

File: tests/integration/test_integration_ml_pipeline.py [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+import cf
+import numpy as np
+
+from tctrack.machine_learning.cyclone_track_ml import (
+    MLParameters,
+    MLStitchParameters,
+    MLTracker,
+)
+
+RDS_DIR = "/home/sg2147/rds/rds-inspire-tc-TqEGHMWTn8A/sg2147"
```
Comment (sjavis) [inline tests/integration/test_integration_ml_pipeline.py:39]:
```suggestion
RDS_DIR = "/rds/project/rds-TqEGHMWTn8A/sg2147"
```

Longer term it would be best if we could make this not need data on CSD3

File: tests/integration/test_integration_ml_pipeline.py [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+    param_overrides = {
+        name: value
+        for name, value in overrides.items()
+        if name not in stitch_field_names
+    }
+    return MLTracker(
+        MLParameters(
+            input_file=input_file,
+            model_path=MODEL_PATH,
+            normalisation_stats_path=stats_file,
```
Comment (sjavis) [inline tests/integration/test_integration_ml_pipeline.py:87]: Can these arguments be ignored here or are the files different to the defaults?

File: tests/integration/test_integration_ml_pipeline.py [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+
+def fetch_norm_stats(stats_file: str) -> None:
+    """Pull the real normalisation statistics from the reference repo."""
+    git = ["git", "-C", REFERENCE_REPO]
+    subprocess.run([*git, "fetch", "origin"], check=True)  # noqa: S603
+    with open(stats_file, "wb") as handle:
+        subprocess.run(  # noqa: S603
+            [*git, "show", REFERENCE_STATS_REF], stdout=handle, check=True
+        )
+    print(f"  fetched statistics from {REFERENCE_STATS_REF}")
```
Comment (sjavis) [inline tests/integration/test_integration_ml_pipeline.py:140]: Should this now be able to use the bundled stats file?

File: tests/integration/test_integration_ml_pipeline.py [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+
+
+def test_pipeline(input_file: str, stats_file: str) -> None:
```
Comment (sjavis) [inline tests/integration/test_integration_ml_pipeline.py:143]: I would suggest converting this and `test_output_writers` to be compatible with pytest. But make it skip them by default with something like:
```
@pytest.mark.skip(reason="Requires data on CSD3")
```

File: tests/integration/test_integration_ml_pipeline.py [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+        "read_trajectories() must return what stitch() produced, else to_netcdf() "
+        "would write nothing"
+    )
+    for trajectory in trajectories:
+        assert trajectory.observations >= tracker.stitch_parameters.stitch_min_length
+        for key in ("time", "lat", "lon"):
+            assert key in trajectory.data, f"trajectory missing '{key}'"
+    print(f"  stitch():     {len(trajectories)} trajectories, all well-formed   OK")
+    if not trajectories:
+        print("                (none found - fine, model quality is not under test)")
```
Comment (sjavis) [inline tests/integration/test_integration_ml_pipeline.py:185]: This should probably check that there is at least one trajectory detected (although I don't know if any would be detected from the input data). It doesn't really check anything otherwise.

**Review by sjavis — CHANGES_REQUESTED** [review]
This is looking good. Thank you for the changes that you previously made. I've left some comments and will open a separate PR for the changes to `Candidate` that I suggested.
